# Data Cleaning of Behavioral Risk Factor Surveillance System
This dataset includes data on adult's diet, physical activity, and weight status from Behavioral Risk Factor Surveillance System. 

Data source: data.gov

# Importing libraries and reading data

In [1]:
import pandas as pd
import numpy as np

In [2]:
df=pd.read_csv('data.csv')
df.head()

YearStart  YearEnd LocationAbbr LocationDesc  \
0       2011     2011           AL      Alabama   
1       2011     2011           AL      Alabama   
2       2011     2011           AL      Alabama   
3       2011     2011           AL      Alabama   
4       2011     2011           AL      Alabama   

                                   Datasource                    Class  \
0  Behavioral Risk Factor Surveillance System  Obesity / Weight Status   
1  Behavioral Risk Factor Surveillance System  Obesity / Weight Status   
2  Behavioral Risk Factor Surveillance System  Obesity / Weight Status   
3  Behavioral Risk Factor Surveillance System  Obesity / Weight Status   
4  Behavioral Risk Factor Surveillance System  Obesity / Weight Status   

                     Topic                                           Question  \
0  Obesity / Weight Status  Percent of adults aged 18 years and older who ...   
1  Obesity / Weight Status  Percent of adults aged 18 years and older who ...   
2  Obesity / Weight Status  Percent of adults aged 18 years and older who ...   
3  Obesity / Weight Status  Percent of adults aged 18 years and older who ...   
4  Obesity / Weight Status  Percent of adults aged 18 years and older who ...   

   Data_Value_Unit Data_Value_Type        ...         \
0              NaN           Value        ...          
1              NaN           Value        ...          
2              NaN           Value        ...          
3              NaN           Value        ...          
4              NaN           Value        ...          

                               GeoLocation  ClassID TopicID QuestionID  \
0  (32.84057112200048, -86.63186076199969)      OWS    OWS1       Q036   
1  (32.84057112200048, -86.63186076199969)      OWS    OWS1       Q036   
2  (32.84057112200048, -86.63186076199969)      OWS    OWS1       Q036   
3  (32.84057112200048, -86.63186076199969)      OWS    OWS1       Q036   
4  (32.84057112200048, -86.63186076199969)      OWS    OWS1       Q036   

   DataValueTypeID  LocationID  StratificationCategory1  \
0            VALUE           1                    Total   
1            VALUE           1                   Gender   
2            VALUE           1                   Gender   
3            VALUE           1                Education   
4            VALUE           1                Education   

         Stratification1 StratificationCategoryId1 StratificationID1  
0                  Total                       OVR           OVERALL  
1                   Male                       GEN              MALE  
2                 Female                       GEN            FEMALE  
3  Less than high school                       EDU             EDUHS  
4   High school graduate                       EDU         EDUHSGRAD  

[5 rows x 33 columns]

In [3]:
df.shape

(48772, 33)

# Finding missing values

In [4]:
df.isnull().sum()

YearStart                         0
YearEnd                           0
LocationAbbr                      0
LocationDesc                      0
Datasource                        0
Class                             0
Topic                             0
Question                          0
Data_Value_Unit               48772
Data_Value_Type                   0
Data_Value                     4603
Data_Value_Alt                 4603
Data_Value_Footnote_Symbol    44169
Data_Value_Footnote           44169
Low_Confidence_Limit           4603
High_Confidence_Limit          4603
Sample_Size                    4603
Total                         47030
Age(years)                    38324
Education                     41804
Gender                        45288
Income                        36578
Race/Ethnicity                34836
GeoLocation                     924
ClassID                           0
TopicID                           0
QuestionID                        0
DataValueTypeID             

# DataValueUnit has all null vaues.So that column can be dropped fully.

In [5]:
df=df.drop('Data_Value_Unit',axis=1)

Data_Value and Data_Value_Alt column has 4603 missing values.Filling that with mean values of that column

In [6]:
df['Data_Value']=df['Data_Value'].fillna(df['Data_Value'].mean())
df['Data_Value'].isnull().sum()

0

In [7]:
df['Data_Value_Alt']=df['Data_Value_Alt'].fillna(df['Data_Value_Alt'].mean())
df['Data_Value_Alt'].isnull().sum()

0

In [8]:
df['Data_Value_Footnote_Symbol'].value_counts()

~    4603
Name: Data_Value_Footnote_Symbol, dtype: int64

In [9]:
df['Data_Value_Footnote'].value_counts()

Data not available because sample size is insufficient.                                                                                                   4181
Data not available because sample size is insufficient.  If data only missing for the confidence interval, the confidence interval was not calculated.     422
Name: Data_Value_Footnote, dtype: int64

Data_Value_Footnote_Symbol and Data_Value_Footnote have more missing values and the other values present doesn't provide much information.So, the two columns are dropped for simplicity of further analysis.

In [10]:
df=df.drop(['Data_Value_Footnote_Symbol','Data_Value_Footnote'],axis=1)

In [11]:
df.isnull().sum()

YearStart                        0
YearEnd                          0
LocationAbbr                     0
LocationDesc                     0
Datasource                       0
Class                            0
Topic                            0
Question                         0
Data_Value_Type                  0
Data_Value                       0
Data_Value_Alt                   0
Low_Confidence_Limit          4603
High_Confidence_Limit         4603
Sample_Size                   4603
Total                        47030
Age(years)                   38324
Education                    41804
Gender                       45288
Income                       36578
Race/Ethnicity               34836
GeoLocation                    924
ClassID                          0
TopicID                          0
QuestionID                       0
DataValueTypeID                  0
LocationID                       0
StratificationCategory1          0
Stratification1                  0
StratificationCatego

Same procedure of filling values for Low_Confidence_Limit,High_Confidence_Limit and Sample_Size

In [12]:
df['Low_Confidence_Limit']=df['Low_Confidence_Limit'].fillna(df['Low_Confidence_Limit'].mean())

In [13]:
df['High_Confidence_Limit ']=df['High_Confidence_Limit '].fillna(df['High_Confidence_Limit '].mean())

In [14]:
df['Sample_Size']=df['Sample_Size'].fillna(df['Sample_Size'].mean())

In [15]:
df.isnull().sum()

YearStart                        0
YearEnd                          0
LocationAbbr                     0
LocationDesc                     0
Datasource                       0
Class                            0
Topic                            0
Question                         0
Data_Value_Type                  0
Data_Value                       0
Data_Value_Alt                   0
Low_Confidence_Limit             0
High_Confidence_Limit            0
Sample_Size                      0
Total                        47030
Age(years)                   38324
Education                    41804
Gender                       45288
Income                       36578
Race/Ethnicity               34836
GeoLocation                    924
ClassID                          0
TopicID                          0
QuestionID                       0
DataValueTypeID                  0
LocationID                       0
StratificationCategory1          0
Stratification1                  0
StratificationCatego

Renaming age column

In [16]:
df=df.rename(columns = {'Age(years)':'Age'})

# Converting Age column to Numeric

In [17]:
def refine_age(age):
    if pd.isnull(age):
        return None
    age=age.split(' ')[0]
    return int(age)
df['Age']=df['Age'].apply(refine_age)
df['Age'].describe()

count    10448.000000
mean        40.499904
std         16.367524
min         18.000000
25%         25.000000
50%         40.000000
75%         55.000000
max         65.000000
Name: Age, dtype: float64

In [18]:
df['Age'].unique()

array([ nan,  18.,  25.,  35.,  45.,  55.,  65.])

# Cleaning the column values of Stratification Category Values

In [19]:
df['Stratification1'].unique()

array(['Total', 'Male', 'Female', 'Less than high school',
       'High school graduate', 'Some college or technical school',
       'College graduate', '18 - 24', '25 - 34', '35 - 44', '45 - 54',
       '55 - 64', '65 or older', 'Less than $15,000', '$15,000 - $24,999',
       '$25,000 - $34,999', '$35,000 - $49,999', '$50,000 - $74,999',
       '$75,000 or greater', 'Data not reported', 'Non-Hispanic White',
       'Non-Hispanic Black', 'Hispanic', 'Asian',
       'Hawaiian/Pacific Islander', 'American Indian/Alaska Native',
       '2 or more races', 'Other'], dtype=object)

In [20]:
df['StratificationCategory1']

0                 Total
1                Gender
2                Gender
3             Education
4             Education
5             Education
6             Education
7           Age (years)
8           Age (years)
9           Age (years)
10          Age (years)
11          Age (years)
12          Age (years)
13               Income
14               Income
15               Income
16               Income
17               Income
18               Income
19               Income
20       Race/Ethnicity
21       Race/Ethnicity
22       Race/Ethnicity
23       Race/Ethnicity
24       Race/Ethnicity
25       Race/Ethnicity
26       Race/Ethnicity
27       Race/Ethnicity
28                Total
29               Gender
              ...      
48742    Race/Ethnicity
48743    Race/Ethnicity
48744       Age (years)
48745       Age (years)
48746       Age (years)
48747       Age (years)
48748       Age (years)
48749       Age (years)
48750         Education
48751         Education
48752         Ed

Observing the Stratification Category1 values, the values in Total,Age,Gender,Education,Race/Ethnicity,Total have values for which the
Stratification Category1 is that respective value.So the other column values are NaN in the table.

So Filling mode or mean of values makes less sense here.Instead it can be filled as 'unknown' and the columns which have values according to the respective Stratification Category1 can be used for analysis.

In [21]:
cols = ['Total',"Age", "Education",'Gender','Income','Race/Ethnicity']
df[cols]=df[cols].fillna('not known')

In [22]:
df.isnull().sum()

YearStart                      0
YearEnd                        0
LocationAbbr                   0
LocationDesc                   0
Datasource                     0
Class                          0
Topic                          0
Question                       0
Data_Value_Type                0
Data_Value                     0
Data_Value_Alt                 0
Low_Confidence_Limit           0
High_Confidence_Limit          0
Sample_Size                    0
Total                          0
Age                            0
Education                      0
Gender                         0
Income                         0
Race/Ethnicity                 0
GeoLocation                  924
ClassID                        0
TopicID                        0
QuestionID                     0
DataValueTypeID                0
LocationID                     0
StratificationCategory1        0
Stratification1                0
StratificationCategoryId1      0
StratificationID1              0
dtype: int

# GeoLocation 

Only the GeoLocation have missing values now.We can try to fill them with Location available in the table.

Before that we can separate the latitude and longitudes in separate columns so that it can be used for analysis and mapping of locations in geomaps.

In [23]:
df['GeoLocation'].head()

0    (32.84057112200048, -86.63186076199969)
1    (32.84057112200048, -86.63186076199969)
2    (32.84057112200048, -86.63186076199969)
3    (32.84057112200048, -86.63186076199969)
4    (32.84057112200048, -86.63186076199969)
Name: GeoLocation, dtype: object

# Extracting Latitude and Longitude values and putting in separate columns

In [24]:
def extract_lat(loc):
    if(pd.isnull(loc)):
        return None
    lat=loc.split(',')[0]
    lat=lat.replace('(','')
    return lat
df['Latitude']=df['GeoLocation'].apply(extract_lat)
df['Latitude'].head()

0    32.84057112200048
1    32.84057112200048
2    32.84057112200048
3    32.84057112200048
4    32.84057112200048
Name: Latitude, dtype: object

In [25]:
def extract_lon(loc):
    if(pd.isnull(loc)):
        return None
    lon=loc.split(',')[1]
    lon=lon.replace(')','')
    return lon
df['Longitude']=df['GeoLocation'].apply(extract_lon)
df['Longitude'].head()

0     -86.63186076199969
1     -86.63186076199969
2     -86.63186076199969
3     -86.63186076199969
4     -86.63186076199969
Name: Longitude, dtype: object

# Filling Missing values with Location name available

In [30]:
import geopy
from geopy.geocoders import GoogleV3

geolocator = GoogleV3() 
null_df=df[df['GeoLocation'].isnull()]
for location in null_df['LocationDesc']:
        geoloc = geolocator.geocode(location)
        null_df['Latitude']=geoloc.latitude
        null_df['Longitude']=geoloc.longitude

C:\Users\harshini\Anaconda3\lib\site-packages\ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
C:\Users\harshini\Anaconda3\lib\site-packages\ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


GeocoderQuotaExceeded: The given key has gone over the requests limit in the 24 hour period or has submitted too many requests in too short a period of time.